### DISCUSSION TITLE
Finding Java Native Memory 'Other' section memory leak and OOM

### DISCUSSION THREAD

**Discussion by adabugra** [async-profiler/async-profiler#1092] [Q&A] [answered]
I am trying to find a memory leak since I get an `java.lang.OutOfMemoryError: Cannot reserve 4194304 bytes of direct buffer memory` after upgrading my minecraft server and plugins. I enabled `NativeMemoryTracking=detail` and noticed that the **Other** section in the `jcmd PID VM.native_memory summary` command consumes 40GB memory starting from 1GB and increasing to the limit until the OOM error occurs:

[nmt-summary.txt](https://github.com/user-attachments/files/18262978/nmt-summary.txt)

I tried nativemem profiling with this:

`start,nativemem=1m,file=malloc.jfr`

and generating flamegraph with these commands:

`./jfrconv --nativemem malloc.jfr malloc.html` -> http://5.180.104.113/malloc.html
`./jfrconv --nativemem --leak malloc.jfr malloc-leak.html` -> http://5.180.104.113/malloc-leak.html

I thought that problem can be related with ProtocolLib plugin (com.comphenix.protocol) and disabled it. However, after a restart, the Other section still consumes 2.5GB of memory and continues to increase.

**Comment by apangin** [comment]
`Other` in Native Memory Tracking summary usually corresponds to the off-heap memory allocated through Direct ByteBuffers or directly via `Unsafe`.

`--nativemem --leak` report confirms that the most unfreed allocations come from `ByteBuffer.allocateDirect` called from Netty's buffer pool:
```
os::malloc
jdk/internal/misc/Unsafe.allocateMemory0
jdk/internal/misc/Unsafe.allocateMemory
java/nio/DirectByteBuffer.<init>
java/nio/ByteBuffer.allocateDirect
io/netty/buffer/PoolArena$DirectArena.allocateDirect
...
```

By default, JDK limits total amount of native memory that can be allocated with direct ByteBuffers to the size of Java Heap. In your case, Java Heap size is 40GB. This means, an application is allowed to allocate 40GB off-heap memory **in addition** to the heap, that is, the total amount of memory consumed by a Java process may exceed 80GB. If you don't have that much memory available, reduce the heap size and/or explicitly set the direct buffer limit with `-XX:MaxDirectMemorySize` JVM option.

Also note that the JVM relies on a Garbage Collector to reclaim direct buffer memory. If an application attempts to allocate more than `MaxDirectMemorySize` before GC occurs, the JVM will attempt to run explicit GC. This will fail if `-XX:+DisableExplicitGC` option is set. Make sure to **remove** `-XX:+DisableExplicitGC` option if it exists.

**Reply by adabugra** [reply]
Removing ViaVersion plugin fixed the memory leak. Flamegraph after memory leak is fixed:
![image](https://github.com/user-attachments/assets/55c672f0-fc15-46fb-99b7-1643235eeb05)

**Reply by apangin** [reply, marked as answer]
Right, `viaversion` is visible in your malloc-leak report, it provokes direct buffer allocation in `com/viaversion/viaversion/util/ByteBufUtil.copy`:
<img width="915" alt="viaversion" src="https://github.com/user-attachments/assets/f7100cca-a33d-4975-ab3b-64ef4d55775a" />

Of course, interpreting the graph requires some knowledge of the code/architecture of the application, and I'm not quite familiar with the Minecraft architecture :)
Thanks for sharing!

**Comment by cloud-66** [comment]
There is native memory leak in my application. I guess byteBuffers spent nativememory, but MNT doesn't show OTHER section, and some additional section from NMT published in this discussion doesnt exists in my NMT.  Is This functionality were added in java versions after jdk 11 which i use ? Thank you.

**Reply by apangin** [reply]
How is the question related to async-profiler?
